internal/core/operations/newschedulerversion/new_scheduler_version_executor.go

```diff
@@ -0,0 +1,143 @@
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package newschedulerversion
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/topfreegames/maestro/internal/core/entities"
+
+	"github.com/Masterminds/semver/v3"
+
+	"github.com/topfreegames/maestro/internal/core/ports/errors"
+
+	"github.com/topfreegames/maestro/internal/core/entities/operation"
+	"github.com/topfreegames/maestro/internal/core/operations"
+	"github.com/topfreegames/maestro/internal/core/services/interfaces"
+	"github.com/topfreegames/maestro/internal/core/services/room_manager"
+	"go.uber.org/zap"
+)
+
+type CreateNewSchedulerVersionExecutor struct {
+	roomManager      *room_manager.RoomManager
+	schedulerManager interfaces.SchedulerManager
+}
+
+func NewExecutor(roomManager *room_manager.RoomManager, schedulerManager interfaces.SchedulerManager) *CreateNewSchedulerVersionExecutor {
+	return &CreateNewSchedulerVersionExecutor{
+		roomManager:      roomManager,
+		schedulerManager: schedulerManager,
+	}
+}
+
+func (ex *CreateNewSchedulerVersionExecutor) Execute(ctx context.Context, op *operation.Operation, definition operations.Definition) error {
+
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", op.DefinitionName),
+		zap.String("operation_id", op.ID),
+	)
+	opDef, ok := definition.(*CreateNewSchedulerVersionDefinition)
+	if !ok {
+		return errors.NewErrInvalidArgument(fmt.Sprintf("invalid operation definition for %s operation", ex.Name()))
+	}
+
+	newScheduler := opDef.NewScheduler
+	currentActiveScheduler, err := ex.schedulerManager.GetActiveScheduler(ctx, opDef.NewScheduler.Name)
+	if err != nil {
+		logger.Error("error getting active scheduler", zap.Error(err))
+		return fmt.Errorf("error getting active scheduler: %w", err)
+	}
+
+	isSchedulerMajorVersion := ex.schedulerManager.IsMajorVersionUpdate(currentActiveScheduler, newScheduler)
+
+	err = ex.populateSchedulerNewVersion(newScheduler, currentActiveScheduler.Spec.Version, isSchedulerMajorVersion)
+	if err != nil {
+		return err
+	}
+
+	if isSchedulerMajorVersion {
+		err = ex.roomManager.ValidateGameRoomCreation(ctx, newScheduler)
+		if err != nil {
+			logger.Error("could not validate new game room creation", zap.Error(err))
+			return err
+		}
+	}
+
+	err = ex.createNewSchedulerVersion(ctx, newScheduler, logger)
+	if err != nil {
+		return err
+	}
+
+	switchActiveVersionOp, err := ex.enqueueSwitchActiveVersionOperation(ctx, newScheduler, logger)
```

Comment:
WDYT about enqueuing the operation with a information like "isMajorChange"? Also, we would have to change on the "switch active version operation" to only switch pods on minor versions (I believe it's changing the pods nevertheless the version change"